File: android/core/designsystem/src/main/kotlin/com/nauhaan/skycast/core/designsystem/theme/WeatherPalette.kt
package com.nauhaan.skycast.core.designsystem.theme

import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import com.nauhaan.skycast.domain.model.WeatherCondition

/**
 * Weather-semantic colours, as a Material 3 **extended colour set**.
 *
 * Colour that encodes *meaning* rather than hierarchy belongs in a custom set alongside the scheme:
 * the `colorScheme` roles all come from one tonal palette, which under Material You is the user's
 * wallpaper, so a sun, a rain cloud and a barometer would render in near-identical tints.
 *
 * Every pair below is a container plus its `on` colour, chosen to clear WCAG AA (4.5:1) for body
 * text at the sizes used here, in both appearances.
 *
 * These stay fixed when Material You is on, so the wallpaper drives the app's chrome but not the
 * weather colours.
 */
@Immutable
data class WeatherPalette(
    val sunContainer: Color,
    val onSunContainer: Color,
    val moonContainer: Color,
    val onMoonContainer: Color,
    val cloudContainer: Color,
    val onCloudContainer: Color,
    val rainContainer: Color,
    val onRainContainer: Color,
    val drizzleContainer: Color,
    val onDrizzleContainer: Color,
    val thunderContainer: Color,
    val onThunderContainer: Color,
    val snowContainer: Color,
    val onSnowContainer: Color,
    val mistContainer: Color,
    val onMistContainer: Color,
    val humidity: Color,
    val wind: Color,
    val pressure: Color,
    val visibility: Color,
    val sunrise: Color,
    val sunset: Color,
    /** Track behind a metric indicator, the unfilled remainder. */
    val metricTrack: Color,
)

/**
 * Light appearance.
 *
 * Containers are soft washes so they sit quietly on a light surface; the `on` colours are deep and
 * saturated so the icon reads as a silhouette rather than a smudge.
 */
val LightWeatherPalette = WeatherPalette(
    sunContainer = Color(0xFFFFE7AE),
    onSunContainer = Color(0xFF7A4E00),
    moonContainer = Color(0xFFDCE1F5),
    onMoonContainer = Color(0xFF2B3768),
    cloudContainer = Color(0xFFE5E9F0),
    onCloudContainer = Color(0xFF3F4A5C),
    rainContainer = Color(0xFFD3E5FB),
    onRainContainer = Color(0xFF0F4478),
    drizzleContainer = Color(0xFFDBEFFA),
    onDrizzleContainer = Color(0xFF115B79),
    thunderContainer = Color(0xFFE8DDFB),
    onThunderContainer = Color(0xFF452A80),
    snowContainer = Color(0xFFE2F2F6),
    onSnowContainer = Color(0xFF0F5165),
    mistContainer = Color(0xFFE4E8E5),
    onMistContainer = Color(0xFF44504B),
    humidity = Color(0xFF1B6EC2),
    wind = Color(0xFF0F7A72),
    pressure = Color(0xFF6A4CA8),
    visibility = Color(0xFF0E7490),
    sunrise = Color(0xFFC9741A),
    sunset = Color(0xFFB4501F),
    metricTrack = Color(0x1A000000),
)

/**
 * Dark appearance.
 *
 * Containers are deep and desaturated so they do not glow against a dark surface; the `on` colours
 * are light tints of the same hue, which keeps each condition recognisable across both themes.
 */
val DarkWeatherPalette = WeatherPalette(
    sunContainer = Color(0xFF4A3812),
    onSunContainer = Color(0xFFFFD98A),
    moonContainer = Color(0xFF232949),
    onMoonContainer = Color(0xFFC2CAEE),
    cloudContainer = Color(0xFF2A303A),
    onCloudContainer = Color(0xFFC8D1E0),
    rainContainer = Color(0xFF113152),
    onRainContainer = Color(0xFFA8CCF2),
    drizzleContainer = Color(0xFF0F3F4E),
    onDrizzleContainer = Color(0xFFA5DBF0),
    thunderContainer = Color(0xFF32225B),
    onThunderContainer = Color(0xFFCCBAF5),
    snowContainer = Color(0xFF113945),
    onSnowContainer = Color(0xFFADDFEB),
    mistContainer = Color(0xFF292F2B),
    onMistContainer = Color(0xFFC2CBC5),
    humidity = Color(0xFF7FB6EE),
    wind = Color(0xFF62C8BF),
    pressure = Color(0xFFB9A2EE),
    visibility = Color(0xFF6FC5DE),
    sunrise = Color(0xFFF0B267),
    sunset = Color(0xFFE58A63),
    metricTrack = Color(0x22FFFFFF),
)

/**
 * The palette in scope.
 *
 * `staticCompositionLocalOf` rather than `compositionLocalOf`: it changes only when the theme
 * flips, so there is no benefit to tracking reads individually and a real cost to doing so.
 */
val LocalWeatherPalette = staticCompositionLocalOf { LightWeatherPalette }

/** Shorthand, mirroring how `MaterialTheme.colorScheme` is read. */
val weatherPalette: WeatherPalette
    @Composable
    @ReadOnlyComposable
    get() = LocalWeatherPalette.current

/**
 * The mood colour of the screen currently being drawn, or `null` where there is no weather
 * background.
 *
 * Provided by `WeatherBackground` around its content. This is the **container** colour, not the mood
 * hue the background wash uses, and the distinction matters: the mood hue is a mid-dark colour by
 * design, cloud's is a slate grey, so laying it over a light card does not tint the card, it dims
 * it. The containers are light pastels in the light theme and deep tones in the dark one, which is
 * what a tinted surface wants in each.
 *
 * `compositionLocalOf`, not `staticCompositionLocalOf`: unlike the palette this changes on every
 * swipe between places, and only the handful of composables that read it should recompose.
 */
val LocalWeatherTint = compositionLocalOf<Color?> { null }

/**
 * The colour a surface on this screen should mix into its fill: the condition's container.
 *
 * See [LocalWeatherTint], which this is not.
 */
@Composable
@ReadOnlyComposable
fun weatherSurfaceTint(condition: WeatherCondition, isDaytime: Boolean): Color = when (condition) {
    WeatherCondition.CLEAR -> if (isDaytime) weatherPalette.sunContainer else weatherPalette.moonContainer
    WeatherCondition.CLOUDS -> if (isDaytime) weatherPalette.cloudContainer else weatherPalette.moonContainer
    WeatherCondition.RAIN -> weatherPalette.rainContainer
    WeatherCondition.DRIZZLE -> weatherPalette.drizzleContainer
    WeatherCondition.THUNDERSTORM -> weatherPalette.thunderContainer
    WeatherCondition.SNOW -> weatherPalette.snowContainer
    WeatherCondition.MIST -> weatherPalette.mistContainer
    WeatherCondition.UNKNOWN -> MaterialTheme.colorScheme.surfaceVariant
}

/**
 * The single hue that carries a condition's mood.
 *
 * Defined here rather than inside the background, which is where it started and where nothing else
 * could reach it, so a warm page could end up with cool cards on it.
 */
@Composable
@ReadOnlyComposable
fun weatherTint(condition: WeatherCondition, isDaytime: Boolean): Color = when (condition) {
    WeatherCondition.CLEAR -> if (isDaytime) weatherPalette.sunrise else weatherPalette.onMoonContainer
    WeatherCondition.CLOUDS -> if (isDaytime) weatherPalette.onCloudContainer else weatherPalette.onMoonContainer
    WeatherCondition.RAIN -> weatherPalette.humidity
    WeatherCondition.DRIZZLE -> weatherPalette.visibility
    WeatherCondition.THUNDERSTORM -> weatherPalette.pressure
    WeatherCondition.SNOW -> weatherPalette.onSnowContainer
    WeatherCondition.MIST -> weatherPalette.onMistContainer
    WeatherCondition.UNKNOWN -> MaterialTheme.colorScheme.surfaceVariant
}

File: ios/SkyCast/Core/DesignSystem/WeatherPalette.swift
import SwiftUI

/// Weather-semantic colours: a hue per condition, so the colour carries meaning rather than
/// hierarchy.
///
/// SF Symbol `.multicolor` rendering fails contrast, since `cloud.fill`, `cloud.moon.fill` and
/// `moon.stars.fill` render white and are invisible on a light surface. `.hierarchical` tinted with
/// one accent makes every condition the same colour.
///
/// Each pair here is a container plus the colour drawn on it, chosen to clear WCAG AA at the sizes
/// used here in **both** appearances. `Color(light:dark:)` resolves per trait collection rather
/// than per launch, so a mid-session appearance change is picked up.
///
/// The Android counterpart is `core/designsystem/theme/WeatherPalette.kt`, with the same hues.
enum WeatherPalette {
    // MARK: - Conditions

    static let sunContainer = Color(light: 0xFFE7AE, dark: 0x4A3812)
    static let onSunContainer = Color(light: 0x7A4E00, dark: 0xFFD98A)
    static let moonContainer = Color(light: 0xDCE1F5, dark: 0x232949)
    static let onMoonContainer = Color(light: 0x2B3768, dark: 0xC2CAEE)
    static let cloudContainer = Color(light: 0xE5E9F0, dark: 0x2A303A)
    static let onCloudContainer = Color(light: 0x3F4A5C, dark: 0xC8D1E0)
    static let rainContainer = Color(light: 0xD3E5FB, dark: 0x113152)
    static let onRainContainer = Color(light: 0x0F4478, dark: 0xA8CCF2)
    static let drizzleContainer = Color(light: 0xDBEFFA, dark: 0x0F3F4E)
    static let onDrizzleContainer = Color(light: 0x115B79, dark: 0xA5DBF0)
    static let thunderContainer = Color(light: 0xE8DDFB, dark: 0x32225B)
    static let onThunderContainer = Color(light: 0x452A80, dark: 0xCCBAF5)
    static let snowContainer = Color(light: 0xE2F2F6, dark: 0x113945)
    static let onSnowContainer = Color(light: 0x0F5165, dark: 0xADDFEB)
    static let mistContainer = Color(light: 0xE4E8E5, dark: 0x292F2B)
    static let onMistContainer = Color(light: 0x44504B, dark: 0xC2CBC5)

    // MARK: - Metrics

    static let humidity = Color(light: 0x1B6EC2, dark: 0x7FB6EE)
    static let wind = Color(light: 0x0F7A72, dark: 0x62C8BF)
    static let pressure = Color(light: 0x6A4CA8, dark: 0xB9A2EE)
    static let visibility = Color(light: 0x0E7490, dark: 0x6FC5DE)
    static let sunrise = Color(light: 0xC9741A, dark: 0xF0B267)
    static let sunset = Color(light: 0xB4501F, dark: 0xE58A63)

    // MARK: - Lookup

    /// Container and content colours for a condition.
    ///
    /// Clear and cloudy skies read differently at night, which is the same distinction the icon
    /// itself makes, so the colour must not stay warm after dark.
    static func colours(for condition: WeatherCondition, isDaytime: Bool) -> (container: Color, content: Color) {
        switch condition {
        case .clear:
            isDaytime ? (sunContainer, onSunContainer) : (moonContainer, onMoonContainer)
        case .clouds:
            isDaytime ? (cloudContainer, onCloudContainer) : (moonContainer, onMoonContainer)
        case .rain: (rainContainer, onRainContainer)
        case .drizzle: (drizzleContainer, onDrizzleContainer)
        case .thunderstorm: (thunderContainer, onThunderContainer)
        case .snow: (snowContainer, onSnowContainer)
        case .mist: (mistContainer, onMistContainer)
        case .unknown: (Color(.secondarySystemFill), Color.secondary)
        }
    }
}

extension Color {
    /// Builds a colour from two hex values, resolved against the current appearance.
    ///
    /// `UIColor(dynamicProvider:)` rather than two static colours picked once: it re-resolves
    /// whenever the trait collection changes, so switching to dark mode mid-session updates
    /// everything without a relaunch.
    init(light: UInt32, dark: UInt32) {
        self.init(uiColor: UIColor { traits in
            UIColor(hex: traits.userInterfaceStyle == .dark ? dark : light)
        })
    }
}

private extension UIColor {
    convenience init(hex: UInt32) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: 1
        )
    }
}
